charmstore/fakeclient.go

```diff
@@ -0,0 +1,582 @@
+
+package charmstore
+
+import (
+	"bytes"
+	"crypto/sha512"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/api/charms"
+	"github.com/juju/juju/core/model"
+	"gopkg.in/juju/charm.v6"
+	charmresource "gopkg.in/juju/charm.v6/resource"
+	"gopkg.in/juju/charmrepo.v2"
+	"gopkg.in/juju/charmrepo.v3/csclient"
+	"gopkg.in/juju/charmrepo.v3/csclient/params"
+	"gopkg.in/macaroon.v1"
+)
+
+// fakeclient.go provides a charmstore client that does not use any HTTP connections
+// or authentication mechanisms
+
+type miniHTTP interface {
+	Get(path string, value interface{}) error
+	Put(path string, value interface{}) error
+}
+
+type CharmStorer interface {
+	miniHTTP
+	UploadCharm(id *charm.URL, ch charm.Charm) (*charm.URL, error)
+	UploadCharmWithRevision(id *charm.URL, ch charm.Charm, promulgatedRevision int) error
+	UploadBundle(id *charm.URL, b charm.Bundle) (*charm.URL, error)
+	UploadBundleWithRevision(id *charm.URL, b charm.Bundle, promulgatedRevision int) error
+	UploadResource(id *charm.URL, name, path string, file io.ReadSeeker, size int64, progress csclient.Progress) (revision int, err error)
+	Publish(id *charm.URL, channels []params.Channel, resources map[string]int) error
+	AddDockerResource(id *charm.URL, resourceName string, imageName, digest string) (revision int, err error)
+	Latest(ids []*charm.URL, headers map[string][]string) ([]params.CharmRevision, error)
+	ListResources(id *charm.URL) ([]params.Resource, error)
+	GetResource(id *charm.URL, name string, revision int) (csclient.ResourceData, error)
+	ResourceMeta(id *charm.URL, name string, revision int) (params.Resource, error)
+	WithChannel(channel params.Channel) CharmStorer
+}
+
+// type ChannelAwareCharmStorer interface {
+// }
+
+// CharmFetcher provides access to charms and other objects
+// held in a charmstore (or locally), such as bundles and resources.
+//
+// c.f. github.com/juju/charmrepo.Interface and cmd/juju/application.DeployAPI
+type CharmFetcher interface {
+	// Get returns the charm referenced by charmURL.
+	Get(charmURL *charm.URL) (charm.Charm, error)
+
+	// GetBundle returns the bundle referenced by bundleURL.
+	GetBundle(bundleURL *charm.URL) (charm.Bundle, error)
+
+	// // GetResource returns the resource referenced by resourceURL.
+	// GetResource(resourceURL *charm.URL, name string, revision int) (csclient.ResourceData, error)
+
+	// ResolveWithChannel resolves an entity to its canonical form, providing the series that
+	// the entity supports as a slice of strings.
+	Resolve(entityURL *charm.URL) (canonicalURL *charm.URL, supportedSeries []string, err error)
+
+	// ResolveWithChannel resolves an entity to its canonical form, providing the series that
+	// the entity supports as a slice of strings. Unlike Resolve, ResolveWithChannel includes the
+	// entity's current channel in the result.
+	ResolveWithChannel(*charm.URL) (*charm.URL, params.Channel, []string, error)
+}
+
+// NoopProgress implements csclient.Progress
```

Comment:
Is `csclient.Progress` an interface? If so, you don't really need to export the type `NoopProgress`. It could just be `noopProgress` right?

Reply:
Yes, it is an interface, but I've ended up removing this struct as we're no longer defining global interfaces within this file, so don't need something that implements `csclient.Progress`.